File: examples/java/RealWorldExample.java
import com.pnf.*;
import java.io.*;
import java.util.*;

public class RealWorldExample {

    static class OHLCBar {
        String timestamp;
        double open;
        double high;
        double low;
        double close;
    }

    static List<OHLCBar> loadCSV(String filename) throws IOException {
        List<OHLCBar> data = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line = br.readLine();

            while ((line = br.readLine()) != null) {
                String[] fields = line.split(",");
                if (fields.length >= 6) {
                    OHLCBar bar = new OHLCBar();
                    bar.timestamp = fields[0];
                    bar.open = Double.parseDouble(fields[2]);
                    bar.high = Double.parseDouble(fields[3]);
                    bar.low = Double.parseDouble(fields[4]);
                    bar.close = Double.parseDouble(fields[5]);
                    data.add(bar);
                }
            }
        }

        System.out.println("Loaded " + data.size() + " bars from " + filename);
        return data;
    }

    static void processDataset(String csvFile, String outputPrefix) throws IOException {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Processing: " + csvFile);
        System.out.println("=".repeat(60));

        List<OHLCBar> data = loadCSV(csvFile);
        if (data.isEmpty()) {
            System.err.println("No data loaded, skipping...");
            return;
        }

        ChartConfig.Builder configBuilder = ChartConfig.builder()
            .method(ConstructionMethod.HIGH_LOW)
            .reversal(3);

        if (csvFile.contains("GBPUSD") || csvFile.contains("EURUSD")) {
            configBuilder.boxSizeMethod(BoxSizeMethod.POINTS).boxSize(0.0020);
        } else {
            configBuilder.boxSizeMethod(BoxSizeMethod.TRADITIONAL).boxSize(0.0);
        }

        ChartConfig config = configBuilder.build();

        Chart chart = new Chart(config);

        long timestamp = System.currentTimeMillis();
        for (OHLCBar bar : data) {
            chart.addData(bar.high, bar.low, bar.close, timestamp);
            timestamp += 1000;
        }

        System.out.println("\nChart Statistics:");
        System.out.println("  Total columns: " + chart.columnCount());
        System.out.println("  X columns: " + chart.xColumnCount());
        System.out.println("  O columns: " + chart.oColumnCount());
        System.out.println("  Box size: " + chart.boxSize());
        System.out.println("  Has bullish bias: " + (chart.hasBullishBias() ? "Yes" : "No"));
        System.out.println("  Has bearish bias: " + (chart.hasBearishBias() ? "Yes" : "No"));

        IndicatorConfig indConfig = IndicatorConfig.builder()
            .smaShortPeriod(5)
            .smaMediumPeriod(10)
            .smaLongPeriod(20)
            .bollingerPeriod(20)
            .bollingerStdDevs(2.0)
            .rsiPeriod(14)
            .rsiOverbought(70.0)
            .rsiOversold(30.0)
            .build();

        Indicators indicators = new Indicators(indConfig);
        indicators.calculate(chart);

        System.out.println("\nIndicator Results:");
        System.out.println("  Bullish Percent: " + String.format("%.2f", indicators.bullishPercent()) + "%");
        System.out.println("  Signals: " + indicators.signalCount());
        System.out.println("  Patterns: " + indicators.patternCount());
        System.out.println("  Buy signals: " + indicators.buySignalCount());
        System.out.println("  Sell signals: " + indicators.sellSignalCount());
        System.out.println("  Support levels: " + indicators.supportLevelCount());
        System.out.println("  Resistance levels: " + indicators.resistanceLevelCount());

        Signal[] signals = indicators.signals();
        System.out.println("\nRecent Signals:");
        for (int i = 0; i < Math.min(5, signals.length); i++) {
            String signalType = signals[i].type() == SignalType.BUY ? "BUY" : "SELL";
            System.out.println("  Column " + signals[i].columnIndex() + ": " + signalType +
                             " (Price: " + String.format("%.5f", signals[i].price()) + ")");
        }

        Pattern[] patterns = indicators.patterns();
        System.out.println("\nRecent Patterns:");
        for (int i = 0; i < Math.min(5, patterns.length); i++) {
            String sentiment = patterns[i].isBullish() ? "BULLISH" : "BEARISH";
            System.out.println("  Column " + patterns[i].columnIndex() + ": " +
                             patterns[i].type() + " (" + sentiment + ", Price: " +
                             String.format("%.5f", patterns[i].price()) + ")");
        }

        System.out.println("\nProcessing complete!");

        chart.close();
        indicators.close();
    }

    public static void main(String[] args) {
        System.out.println("PnF Chart Library - Real World Data Example (Java)");

        try {
            processDataset("../../fixtures/GBPUSD_PERIOD_M1.csv", "output/gbpusd_m1");
            processDataset("../../fixtures/Boom_500_Index_PERIOD_H1.csv", "output/boom500_h1");
            processDataset("../../fixtures/Volatility_75_Index_PERIOD_D1.csv", "output/volatility75_d1");

            System.out.println("\n" + "=".repeat(60));
            System.out.println("All datasets processed successfully!");
            System.out.println("=".repeat(60));

        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: examples/rust/real_world_example.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Write};
use std::path::Path;

extern crate pnf;
use pnf::*;

#[derive(Debug)]
struct OHLCBar {
    timestamp: String,
    open: f64,
    high: f64,
    low: f64,
    close: f64,
}

fn load_csv(filename: &str) -> Result<Vec<OHLCBar>, Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut data = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        if i == 0 {
            continue;
        }

        let line = line?;
        let fields: Vec<&str> = line.split(',').collect();

        if fields.len() >= 6 {
            let bar = OHLCBar {
                timestamp: fields[0].to_string(),
                open: fields[2].parse()?,
                high: fields[3].parse()?,
                low: fields[4].parse()?,
                close: fields[5].parse()?,
            };
            data.push(bar);
        }
    }

    println!("Loaded {} bars from {}", data.len(), filename);
    Ok(data)
}

fn process_dataset(csv_file: &str, output_prefix: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("\n{}", "=".repeat(60));
    println!("Processing: {}", csv_file);
    println!("{}", "=".repeat(60));

    let data = load_csv(csv_file)?;
    if data.is_empty() {
        eprintln!("No data loaded, skipping...");
        return Ok(());
    }

    let (box_size_method, box_size) = if csv_file.contains("GBPUSD") || csv_file.contains("EURUSD") {
        (BoxSizeMethod::Points, 0.0020)
    } else {
        (BoxSizeMethod::Traditional, 0.0)
    };

    let config = ChartConfig {
        method: ConstructionMethod::HighLow,
        box_size_method,
        box_size,
        reversal: 3,
    };

    let mut chart = Chart::with_config(config);

    let mut timestamp: i64 = 1640000000;
    for bar in &data {
        chart.add_data(bar.high, bar.low, bar.close, timestamp);
        timestamp += 1;
    }

    println!("\nChart Statistics:");
    println!("  Total columns: {}", chart.column_count());
    println!("  X columns: {}", chart.x_column_count());
    println!("  O columns: {}", chart.o_column_count());
    println!("  Box size: {}", chart.box_size());
    println!("  Has bullish bias: {}", if chart.has_bullish_bias() { "Yes" } else { "No" });
    println!("  Has bearish bias: {}", if chart.has_bearish_bias() { "Yes" } else { "No" });

    let ind_config = IndicatorConfig {
        sma_short_period: 5,
        sma_medium_period: 10,
        sma_long_period: 20,
        bollinger_period: 20,
        bollinger_std_devs: 2.0,
        rsi_period: 14,
        rsi_overbought: 70.0,
        rsi_oversold: 30.0,
        bullish_alert_threshold: 70.0,
        bearish_alert_threshold: 30.0,
        support_resistance_threshold: 0.01,
        congestion_min_columns: 4,
        congestion_price_range: 0.05,
    };

    let mut indicators = Indicators::with_config(ind_config);
    indicators.calculate(&chart);

    println!("\nIndicator Results:");
    println!("  Bullish Percent: {:.2}%", indicators.bullish_percent());
    println!("  Signals: {}", indicators.signal_count());
    println!("  Patterns: {}", indicators.pattern_count());
    println!("  Buy signals: {}", indicators.buy_signal_count());
    println!("  Sell signals: {}", indicators.sell_signal_count());
    println!("  Support levels: {}", indicators.support_level_count());
    println!("  Resistance levels: {}", indicators.resistance_level_count());

    let signals = indicators.signals();
    println!("\nRecent Signals:");
    for signal in signals.iter().take(5) {
        let signal_type = match signal.signal_type {
            SignalType::Buy => "BUY",
            SignalType::Sell => "SELL",
            SignalType::None => "NONE",
        };
        println!("  Column {}: {} (Price: {:.5})", signal.column_index, signal_type, signal.price);
    }

    let patterns = indicators.patterns();
    println!("\nRecent Patterns:");
    for pattern in patterns.iter().take(5) {
        let sentiment = if pattern.is_bullish { "BULLISH" } else { "BEARISH" };
        println!("  Columns {}-{}: {:?} ({}, Price: {:.5})",
                 pattern.start_column, pattern.end_column,
                 pattern.pattern_type, sentiment, pattern.price);
    }

    println!("\nGenerating visualizations...");

    let output_dir = Path::new(output_prefix).parent().unwrap();
    std::fs::create_dir_all(output_dir)?;

    if let Some(ascii_str) = chart.to_ascii() {
        let mut file = File::create(format!("{}_ascii.txt", output_prefix))?;
        file.write_all(ascii_str.as_bytes())?;
        println!("  ASCII chart saved to: {}_ascii.txt", output_prefix);
    }

    if let Some(json_str) = chart.to_json() {
        let mut file = File::create(format!("{}_chart.json", output_prefix))?;
        file.write_all(json_str.as_bytes())?;
        println!("  JSON chart saved to: {}_chart.json", output_prefix);
    }

    println!("\nProcessing complete!");

    Ok(())
}

fn main() {
    println!("PnF Chart Library - Real World Data Example (Rust)");
    println!("Library Version: {}", pnf::version());

    if let Err(e) = process_dataset("../../fixtures/GBPUSD_PERIOD_M1.csv", "output/gbpusd_m1") {
        eprintln!("Error processing GBPUSD: {}", e);
    }

    if let Err(e) = process_dataset("../../fixtures/Boom_500_Index_PERIOD_H1.csv", "output/boom500_h1") {
        eprintln!("Error processing Boom 500: {}", e);
    }

    if let Err(e) = process_dataset("../../fixtures/Volatility_75_Index_PERIOD_D1.csv", "output/volatility75_d1") {
        eprintln!("Error processing Volatility 75: {}", e);
    }

    println!("\n{}", "=".repeat(60));
    println!("All datasets processed successfully!");
    println!("{}", "=".repeat(60));
}
